From 1419dd0a0bb31d7a783ca3da371526ec8ac647a8 Mon Sep 17 00:00:00 2001
From: David Lamparter <equinox@opensourcerouting.org>
Date: Sun, 11 May 2025 16:38:34 +0200
Subject: [PATCH] lib/clippy: pointer offsets are signed

Fedora 42 has some new GCC/ld combination that has negative offsets from
the .note.FRR to the xref pointers.  (This is completely fine, those
offsets are supposed to be signed.)  Clippy decoded them as unsigned,
resulting in off-by-2^64 offset values (which Python cheerfully
processes, due to its builtin "large integer" support... in C code it
would've just wrapped in an uint64_t and made no difference...)

Read the values as signed like they should be.

Signed-off-by: David Lamparter <equinox@opensourcerouting.org>
---
 python/xrelfo.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/python/xrelfo.py b/python/xrelfo.py
index 5f7616f25093..e70c33e62cb6 100644
--- a/python/xrelfo.py
+++ b/python/xrelfo.py
@@ -385,11 +385,11 @@ def load_elf(self, filename, orig_filename):
             endian = ">" if edf._elffile.bigendian else "<"
             mem = edf._elffile[note]
             if edf._elffile.elfclass == 64:
-                start, end = struct.unpack(endian + "QQ", mem)
+                start, end = struct.unpack(endian + "qq", mem)
                 start += note.start
                 end += note.start + 8
             else:
-                start, end = struct.unpack(endian + "II", mem)
+                start, end = struct.unpack(endian + "ii", mem)
                 start += note.start
                 end += note.start + 4
 
